DEPR: deprecate silent alignment on arithmetic operations between Series with non-aligned MultiIndexes

natmokval opened the pull request:
xref #25891, #19606, #32569
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

natmokval committed:
deprecate alignment arithmetic between Series with non-aligned MultiIndexeså

natmokval committed:
add check if other is Series

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

jbrockmendel commented on pandas/core/series.py:
```diff
@@ -6696,6 +6696,21 @@ def _logical_method(self, other, op):
+                # GH#25891
+                warnings.warn(
+                    "The silent alignment on arithmetic operations between "
+                    "'Series' with non-aligned MultiIndexes will be removed "
```
pls make sure to be explicit "is deprecated and ..."

natmokval committed:
resolve conflicts and merge

natmokval committed:
correct the deprecation condition and message

natmokval committed:
fix error in advanced.rst

natmokval committed:
fix an error in groupby.rst

natmokval committed:
fix test_pivot.py and test_multilevel.py

natmokval committed:
fix error in visualization.rst

natmokval committed:
fix tests/indexing/test_loc.py

natmokval committed:
fix error in v0.21.0.rst

rhshadrach reviewed (changes requested):
I'm -1 on the PR as it currently stands, I left a comment in the linked issue.

Edit: I must have done something wrong; in the example in the comment I thought I was getting a deprecation warning with this PR. Now I'm not seeing it. Looking more closely here.

Edit 2: I see, my original example had two MultiIndexes and warned, and then I simplified to MultiIndex + Index and that doesn't warn. This is inconsistent, and I'm negative on deprecating the `MultiIndex & MultiIndex` case as mentioned.

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

natmokval committed:
add check if MultiIndexes are equal

jbrockmendel commented on pandas/core/groupby/groupby.py:
```diff
@@ -5620,7 +5620,7 @@ def diff(
                 shifted = shifted.astype(dict.fromkeys(to_coerce, "float32"))
 
-        return obj - shifted
+        return obj.sub(shifted)
```
these changes are necessary to avoid warnings?

natmokval committed:
revert unnecessary changes

natmokval committed:
correct the deprecation conditions

natmokval committed:
correct the deprecation conditions to fix test_series_varied_multiindex_alignment

natmokval committed:
revert changes in whatsnew/v0.21.0.rst and user_guide

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

rhshadrach commented on pandas/core/series.py:
```diff
@@ -7009,6 +7009,26 @@ def _logical_method(self, other, op) -> Series:
+                    self.index.equals(other.index)
+                    or self.index.isin(other.index).all()
+                    or other.index.isin(self.index).all()
+                )
```
Mentioned in https://github.com/pandas-dev/pandas/issues/25891#issuecomment-4093164838, I'm only okay with this deprecation when the levels are not comparable, not the values themselves. Can you change this to check when one set of `names` is a subset of the other. This might require some careful thought when there are duplicates.

natmokval replied:
Thanks @rhshadrach for the comment. I changed the condition to check whether one set of names is a subset of the other. Could you please take a look at my changes? Is that what you meant?

natmokval committed:
change the condition to check: is one set of names a subset of the other

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

rhshadrach commented on pandas/core/series.py:
```diff
@@ -7009,6 +7009,41 @@ def _logical_method(self, other, op) -> Series:
+                        is_other_in_self = False
+                        break
+
+                if not (is_self_in_other or is_other_in_self):
```
Just use sets.


```suggestion
                self_names = set(self.index.names)
                other_names = set(other.index.names)
                if not (self_names <= other_names or other_names <= self_names):
```

natmokval replied:
done

rhshadrach commented on pandas/core/series.py:
```diff
@@ -7009,6 +7009,41 @@ def _logical_method(self, other, op) -> Series:
+                    warnings.warn(
+                        "The silent alignment on arithmetic operations between "
+                        "'Series' with non-aligned MultiIndexes is deprecated "
+                        "and will be removed in a future version.",
```
With my preferred behavior, we still support various non-aligned cases.


```suggestion
                        "'Series' with incomparable MultiIndexes is deprecated "
                        "and will be removed in a future version.",
```

rhshadrach commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -770,7 +770,7 @@ def test_gb_apply_list_of_unequal_len_arrays():
 
     def noddy(value, weight):
-        out = np.array(value * weight).repeat(3)
+        out = np.array(value.mul(weight)).repeat(3)
```
I suspect many of the test changes can now be reverted. Can you check.

rhshadrach commented on pandas/tests/series/test_arithmetic.py:
```diff
@@ -1115,3 +1115,36 @@ def test_comparison_mismatched_datetime_units(index):
     tm.assert_series_equal(result2, expected2)
+
+
+def test_alignment_on_arithmetic_with_nonaligned_multiindex_deprecated(
```
```suggestion
def test_arithmetic_with_incomparible_multiindex_deprecated(
```

rhshadrach reviewed (changes requested):
Looking good! Sorry this fell off my radar; some requets.

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

natmokval committed:
revert changes in tests

natmokval committed:
update the note in whatsnew

natmokval committed:
correct message in test_arithmetic_with_incomparible_multiindex_deprecated

natmokval commented:
Thanks @rhshadrach for reviewing my changes. I did as you suggested and updated my PR.

rhshadrach commented on doc/source/user_guide/advanced.rst:
```diff
@@ -210,6 +210,7 @@ axes will work as you expect; data alignment will work the same as an Index of
 tuples:
 
 .. ipython:: python
+   :okwarning:
```
Why is this needed?

natmokval replied:
This was a leftover from previous changes. It is no longer needed, so I reverted it.

rhshadrach commented on pandas/tests/series/test_arithmetic.py:
```diff
@@ -1115,3 +1115,36 @@ def test_comparison_mismatched_datetime_units(index):
+):
+    # GH#25891
+    index1 = pd.MultiIndex.from_product([[], []], names=["T", "N"])
+    s1 = Series(index=index1)
+    s1["T.1A", "N.0"] = 0.5
+    s1["T.1B", "N.0"] = 0.5
+    op = tm.get_op_from_name(all_arithmetic_operators)
+
+    index2 = pd.MultiIndex.from_product([[], []], names=["N", "M"])
+    s2 = Series(index=index2)
+    s2["N.0", "M.0"] = 0.5
+
+    expected = Series(
+        [0.25, 0.25],
+        index=pd.MultiIndex.from_tuples(
+            [("T.1A", "N.0", "M.0"), ("T.1B", "N.0", "M.0")],
+            names=["T", "N", "M"],
+        ),
+    )
+    msg = (
+        "The silent alignment on arithmetic operations between "
+        "'Series' with incomparable MultiIndexes"
+    )
+    with tm.assert_produces_warning(Pandas4Warning, match=msg):
+        result = s1 * s2
+    tm.assert_series_equal(result, expected)
+
+    with tm.assert_produces_warning(Pandas4Warning, match=msg):
+        op(s1, s2)
```
Why do both of these? Isn't multiplication one of the ops?

natmokval replied:
Yes, that’s right. I added the multiplication case to compare `expected` and `result`, but it is cleaner to keep only one example that checks the warning message. I deleted the multiplication case.

rhshadrach replied:
Ahh, I see! Yes, I think it's okay to not check the result in this case.

natmokval committed:
update advanced.rst

natmokval committed:
simlify test_arithmetic_with_incomparible_multiindex_deprecated

rhshadrach reviewed (dismissed):
lgtm

rhshadrach commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -956,6 +956,7 @@ Other Deprecations
 - Deprecated backward-compatibility behavior for :meth:`DataFrame.select_dtypes` matching ``str`` dtype when ``np.object_`` is specified (:issue:`61916`)
 - Deprecated option ``future.no_silent_downcasting``, as it is no longer used. In a future version accessing this option will raise (:issue:`59502`)
 - Deprecated passing non-Index types to :meth:`Index.join`; explicitly convert to Index first (:issue:`62897`)
+- Deprecated silent alignment on arithmetic operations between :class:`Series` with incomparable ``MultiIndexes`` (:issue:`25891`)
```
Will need to be moved to 3.1.0 now.

natmokval replied:
thanks, it's done.

rhshadrach commented:
@jbrockmendel - you good here?

natmokval committed:
move the whatsnew note to v3.1.0.rst

jbrockmendel commented on pandas/core/series.py:
```diff
@@ -7045,6 +7045,22 @@ def _logical_method(self, other, op) -> Series:
+                    warnings.warn(
+                        "The silent alignment on arithmetic operations between "
+                        "'Series' with incomparable MultiIndexes is deprecated "
+                        "and will be removed in a future version.",
```
is there a "Do X instead"?  I guess align manually?

jbrockmendel replied:
Is there an analogous DataFrame operation we want to deprecate?

natmokval replied:
﻿﻿Thanks, I added “Please align MultiIndexes manually.” to the warning message.

natmokval replied:
If we append two DataFrames with the same MultiIndexes as in the `test_arithmetic_with_incomparible_multiindex_deprecated` that checks that the expected deprecation warning message is raised for two Series, we won’t see the warning:

```
index1 = pd.MultiIndex.from_product([[], []], names=["T", "N"])
index2 = pd.MultiIndex.from_product([[], []], names=["N", "M"])

df1 = pd.DataFrame(index=index1, columns=["value"])
df2 = pd.DataFrame(index=index2, columns=["value"])

df1.loc[("T.1A", "N.0"), "value"] = 0.5
df1.loc[("T.1B", "N.0"), "value"] = 0.5

df2.loc[("N.0", "M.0"), "value"] = 0.5

print(df1 + df2)
              value
# T    N   M        
# T.1A N.0 M.0   1.0
# T.1B N.0 M.0   1.0
```
I suggest raising the same deprecation warning in this case.

jbrockmendel replied:
> I suggest raising the same deprecation warning in this case.

Yes I think we should keep the behavior consistent.

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

natmokval committed:
correct the warning message

jbrockmendel commented:
> and `series1.add(series2)` will not.

Did you want to address this for the flex ops too?

jbrockmendel commented on pandas/tests/series/test_arithmetic.py:
```diff
@@ -1115,3 +1115,26 @@ def test_comparison_mismatched_datetime_units(index):
     tm.assert_series_equal(result2, expected2)
+
+
+def test_arithmetic_with_incomparible_multiindex_deprecated(
```
incomparible -> incomparable

natmokval replied:
Done

jbrockmendel commented:
Does this merit a line in the docs?

natmokval committed:
Merge branch 'main' into depr-alignment-arithmetic-between-series-with-non-aligned-MultiIndexes

natmokval committed:
fix typo

natmokval commented:
> > and `series1.add(series2)` will not.
> 
> Did you want to address this for the flex ops too?

Yes, I think we should raise the same warning message for operations like add, mul, div, etc. I will open a separate PR for these changes.

natmokval commented:
> Does this merit a line in the docs?

Not yet. I am going to open a separate PR to deprecate silent alignment between Series/DataFrame with non-aligned MultiIndexes for operations like add, mul, div, etc. There I can clarify the new behavior in the docs.

jbrockmendel commented:
> I am going to open a separate PR to deprecate silent alignment between Series/DataFrame with non-aligned MultiIndexes for operations like add, mul, div, etc. There I can clarify the new behavior in the docs.

I know we usually ask for bite-sized PRs, but in this case can we do it all at once? I'm planning to bring the API discussion up in the next dev call, and I'd like to be able to point at a unified "this is the suggested API"

natmokval committed:
resolve conflicts

natmokval committed:
add deprecation warning when silent alignmentwith non-aligned MultiIndexes for operations like add, mul, div, etc.

natmokval committed:
remove duplicate test

natmokval committed:
add test_flex_method_with_incomparable_multiindex_deprecated

natmokval committed:
update docs for Series arithmetic operations

natmokval committed:
deprecate alignment arithmetic/flex ops between DataFrame with incomparable MultiIndexes, add tests

natmokval committed:
update v3.1.0 and docs for DataFrame arithmetic operations

natmokval committed:
fix docs

natmokval committed:
remove extra blank lines from docs

natmokval committed:
add missing line to DataFrame.rtruediv docs

natmokval commented:
> > I am going to open a separate PR to deprecate silent alignment between Series/DataFrame with non-aligned MultiIndexes for operations like add, mul, div, etc. There I can clarify the new behavior in the docs.
> 
> I know we usually ask for bite-sized PRs, but in this case can we do it all at once? I'm planning to bring the API discussion up in the next dev call, and I'd like to be able to point at a unified "this is the suggested API"

Thanks for the comment. I deprecated silent alignment between Series with incomparable MultiIndexes for flex ops, as well as silent alignment between DataFrames with incomparable MultiIndexes for arithmetic and flex operations. I also added a note to the docs that this behavior is deprecated. CI is green.

natmokval commented:
gentle ping @jbrockmendel

jbrockmendel commented:
Discussed in last week's dev call.  I think the API we landed on was to allow non-subset names, treating it like a merge on shared levels.  @rhshadrach can you confirm

rhshadrach commented:
Yes - agreed!

natmokval committed:
changethe condition and check if the MultiIndexes have no common names

natmokval commented:
> Discussed in last week's dev call. I think the API we landed on was to allow non-subset names, treating it like a merge on shared levels. @rhshadrach can you confirm

thanks, @jbrockmendel . If I understand correctly, we still deprecate the silent alignment on arithmetic operations between `Series/DataFrame` with incomparable MultiIndexes when the MultiIndexes have no common names. I changed the condition to allow non-subset names and updated my PR.

natmokval commented:
gentle ping @jbrockmendel

I think, I adressed all comments. Could you please take a look?

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -9624,6 +9624,23 @@ def _cmp_method(self, other, op) -> DataFrame:
+            ):
+                self_names = set(self.index.names)
+                other_names = set(other.index.names)
+                if self.index.equals(other.index) and not (self_names & other_names):
```
can we do the names check first?  i think the index.equals check can be relatively expensive

natmokval replied:
Thanks for the comment. Done

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -9624,6 +9624,23 @@ def _cmp_method(self, other, op) -> DataFrame:
         return self._construct_result(new_data, other=other)
 
     def _arith_method(self, other, op) -> DataFrame:
```
I think the cases we need to think about are 3x2x2

[arith_method, cmp_method, logical_method]
[Series, DataFrame]
[flex, not-flex]

Do they all get the same treatment?  e.g.

```
mi1 = pd.MultiIndex.from_product([["a"], [1]], names=["x", "y"])
mi2 = pd.MultiIndex.from_product([["a"], [1]], names=["p", "q"])
s1 = pd.Series([1], index=mi1)
s2 = pd.Series([1], index=mi2)
df1 = pd.DataFrame(s1)
df2 = pd.DataFrame(s2)

s1.eq(s2)
s1 == s2
df1.eq(df2)
df1 == df2
[...]
```

natmokval replied:
Thanks, you are right, there are some inconsistencies.

For example, for comparison operations, we raise the warning only for `s1.eq(s2)`, while `df1.eq(df2)` and the `== `op for both Series and DataFrame work silently. I will work on fixing these inconsistencies.
